File: Go/src/modbus/data_storage.go
package main

import (
	"database/sql"
	"encoding/csv"
	"fmt"
	"os"
	"sync"
	"time"

	influxdb2 "github.com/influxdata/influxdb-client-go/v2"
	"github.com/influxdata/influxdb-client-go/v2/api"
	_ "github.com/mattn/go-sqlite3"
)

const (
	StorageTypeSQLite   = "sqlite"
	StorageTypeInfluxDB = "influxdb"
	StorageTypeCSV      = "csv"
	StorageTypeNone     = "none"
)

type SensorRecord struct {
	SensorName  string
	SlaveID     int
	Temperature float64
	Humidity    float64
	Timestamp   time.Time
}

type DataStorage interface {
	Save(record *SensorRecord) error
	SaveBatch(records []*SensorRecord) error
	Close() error
}

type SQLiteStorage struct {
	dbPath string
	db     *sql.DB
}

func NewSQLiteStorage(dbPath string) (*SQLiteStorage, error) {
	db, err := sql.Open("sqlite3", dbPath)
	if err != nil {
		return nil, fmt.Errorf("打开数据库失败: %v", err)
	}

	_, err = db.Exec(`
		CREATE TABLE IF NOT EXISTS sensor_data (
			id INTEGER PRIMARY KEY AUTOINCREMENT,
			timestamp DATETIME DEFAULT CURRENT_TIMESTAMP,
			sensor_name TEXT NOT NULL,
			slave_id INTEGER,
			temperature REAL,
			humidity REAL
		)
	`)
	if err != nil {
		return nil, fmt.Errorf("创建表失败: %v", err)
	}

	_, err = db.Exec("CREATE INDEX IF NOT EXISTS idx_timestamp ON sensor_data(timestamp)")
	if err != nil {
		return nil, fmt.Errorf("创建索引失败: %v", err)
	}

	_, err = db.Exec("CREATE INDEX IF NOT EXISTS idx_sensor_name ON sensor_data(sensor_name)")
	if err != nil {
		return nil, fmt.Errorf("创建索引失败: %v", err)
	}

	return &SQLiteStorage{dbPath: dbPath, db: db}, nil
}

func (s *SQLiteStorage) Save(record *SensorRecord) error {
	_, err := s.db.Exec(
		"INSERT INTO sensor_data (sensor_name, slave_id, temperature, humidity, timestamp) VALUES (?, ?, ?, ?, ?)",
		record.SensorName, record.SlaveID, record.Temperature, record.Humidity, record.Timestamp,
	)
	return err
}

func (s *SQLiteStorage) SaveBatch(records []*SensorRecord) error {
	tx, err := s.db.Begin()
	if err != nil {
		return err
	}
	defer tx.Rollback()

	stmt, err := tx.Prepare(
		"INSERT INTO sensor_data (sensor_name, slave_id, temperature, humidity, timestamp) VALUES (?, ?, ?, ?, ?)",
	)
	if err != nil {
		return err
	}
	defer stmt.Close()

	for _, record := range records {
		_, err := stmt.Exec(
			record.SensorName, record.SlaveID, record.Temperature, record.Humidity, record.Timestamp,
		)
		if err != nil {
			return err
		}
	}

	return tx.Commit()
}

func (s *SQLiteStorage) Query(limit, offset int) ([]map[string]interface{}, error) {
	rows, err := s.db.Query(
		"SELECT * FROM sensor_data ORDER BY timestamp DESC LIMIT ? OFFSET ?",
		limit, offset,
	)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	columns, err := rows.Columns()
	if err != nil {
		return nil, err
	}

	var results []map[string]interface{}
	for rows.Next() {
		row := make(map[string]interface{})
		values := make([]interface{}, len(columns))
		valuePtrs := make([]interface{}, len(columns))
		for i := range values {
			valuePtrs[i] = &values[i]
		}

		err := rows.Scan(valuePtrs...)
		if err != nil {
			return nil, err
		}

		for i, col := range columns {
			row[col] = values[i]
		}
		results = append(results, row)
	}

	return results, nil
}

func (s *SQLiteStorage) GetStats() (map[string]interface{}, error) {
	var count int
	err := s.db.QueryRow("SELECT COUNT(*) FROM sensor_data").Scan(&count)
	if err != nil {
		return nil, err
	}
	return map[string]interface{}{"total_records": count}, nil
}

func (s *SQLiteStorage) Close() error {
	return s.db.Close()
}

type InfluxDBStorage struct {
	url      string
	token    string
	org      string
	bucket   string
	client   influxdb2.Client
	writeAPI api.WriteAPI
}

func (s *InfluxDBStorage) Save(record *SensorRecord) error {
	point := influxdb2.NewPoint("sensor_data",
		map[string]string{
			"sensor":   record.SensorName,
			"slave_id": fmt.Sprintf("%d", record.SlaveID),
		},
		map[string]interface{}{
			"temperature": record.Temperature,
			"humidity":    record.Humidity,
		},
		record.Timestamp,
	)
	s.writeAPI.WritePoint(point)
	return nil
}

func (s *InfluxDBStorage) SaveBatch(records []*SensorRecord) error {
	for _, record := range records {
		point := influxdb2.NewPoint("sensor_data",
			map[string]string{
				"sensor":   record.SensorName,
				"slave_id": fmt.Sprintf("%d", record.SlaveID),
			},
			map[string]interface{}{
				"temperature": record.Temperature,
				"humidity":    record.Humidity,
			},
			record.Timestamp,
		)
		s.writeAPI.WritePoint(point)
	}
	return nil
}

func (s *InfluxDBStorage) Close() error {
	s.writeAPI.Flush()
	s.client.Close() // Close doesn't return a value
	return nil
}

type CSVStorage struct {
	filePath string
	file     *os.File
}

func NewCSVStorage(filePath string) (*CSVStorage, error) {
	var file *os.File
	var err error

	if _, err = os.Stat(filePath); os.IsNotExist(err) {
		file, err = os.Create(filePath)
		if err != nil {
			return nil, err
		}
		writer := csv.NewWriter(file)
		writer.Write([]string{"timestamp", "sensor_name", "slave_id", "temperature", "humidity"})
		writer.Flush()
	} else {
		file, err = os.OpenFile(filePath, os.O_APPEND|os.O_WRONLY, 0644)
		if err != nil {
			return nil, err
		}
	}

	return &CSVStorage{filePath: filePath, file: file}, nil
}

func (s *CSVStorage) Save(record *SensorRecord) error {
	writer := csv.NewWriter(s.file)
	defer writer.Flush()

	return writer.Write([]string{
		record.Timestamp.Format("2006-01-02 15:04:05"),
		record.SensorName,
		fmt.Sprintf("%d", record.SlaveID),
		fmt.Sprintf("%.1f", record.Temperature),
		fmt.Sprintf("%.1f", record.Humidity),
	})
}

func (s *CSVStorage) SaveBatch(records []*SensorRecord) error {
	writer := csv.NewWriter(s.file)
	defer writer.Flush()

	for _, record := range records {
		err := writer.Write([]string{
			record.Timestamp.Format("2006-01-02 15:04:05"),
			record.SensorName,
			fmt.Sprintf("%d", record.SlaveID),
			fmt.Sprintf("%.1f", record.Temperature),
			fmt.Sprintf("%.1f", record.Humidity),
		})
		if err != nil {
			return err
		}
	}

	return nil
}

func (s *CSVStorage) Close() error {
	return s.file.Close()
}

type StorageFactory struct{}

func (f *StorageFactory) CreateStorage(config *StorageConfig) (DataStorage, error) {
	var storage DataStorage
	var err error

	switch config.Type {
	case StorageTypeSQLite:
		storage, err = NewSQLiteStorage(config.SQLiteConfig.DBPath)
	case StorageTypeInfluxDB:
		options := influxdb2.DefaultOptions()
		client := influxdb2.NewClientWithOptions(
			config.InfluxDBConfig.URL,
			config.InfluxDBConfig.Token,
			options,
		)
		storage = &InfluxDBStorage{
			url:      config.InfluxDBConfig.URL,
			token:    config.InfluxDBConfig.Token,
			org:      config.InfluxDBConfig.Org,
			bucket:   config.InfluxDBConfig.Bucket,
			client:   client,
			writeAPI: client.WriteAPI(config.InfluxDBConfig.Org, config.InfluxDBConfig.Bucket),
		}
	case StorageTypeCSV:
		storage, err = NewCSVStorage(config.CSVConfig.FilePath)
	case StorageTypeNone:
		return nil, nil
	default:
		storage, err = NewSQLiteStorage(config.SQLiteConfig.DBPath)
	}

	if err != nil {
		return nil, err
	}

	if storage == nil {
		return nil, nil
	}

	if config.CacheConfig.Enabled {
		cacheDBPath := config.CacheConfig.DBPath
		if cacheDBPath == "" {
			cacheDBPath = "sensor_cache.db"
		}
		flushInterval := config.CacheConfig.FlushInterval
		if flushInterval <= 0 {
			flushInterval = 10
		}
		batchSize := config.CacheConfig.BatchSize
		if batchSize <= 0 {
			batchSize = 100
		}

		return NewCachedStorage(storage, cacheDBPath, flushInterval, batchSize)
	}

	return storage, nil
}

func GetDefaultStorageConfig() *StorageConfig {
	return &StorageConfig{
		Type:           StorageTypeSQLite,
		SQLiteConfig:   SQLiteConfig{DBPath: "sensor_data.db"},
		InfluxDBConfig: InfluxDBConfig{URL: "http://localhost:8086", Token: "my-token", Org: "my-org", Bucket: "sensor-data"},
		CSVConfig:      CSVConfig{FilePath: "sensor_data.csv"},
		CacheConfig:    CacheConfig{Enabled: false, DBPath: "sensor_cache.db", FlushInterval: 10, BatchSize: 100},
	}
}

type CacheRecord struct {
	ID          int64
	SensorName  string
	SlaveID     int
	Temperature float64
	Humidity    float64
	Timestamp   time.Time
	RetryCount  int
}

type CachedStorage struct {
	targetStorage DataStorage
	cacheDBPath   string
	cacheDB       *sql.DB
	flushInterval int
	batchSize     int
	running       bool
	mu            sync.Mutex
}

func NewCachedStorage(targetStorage DataStorage, cacheDBPath string, flushInterval, batchSize int) (*CachedStorage, error) {
	cacheDB, err := sql.Open("sqlite3", cacheDBPath)
	if err != nil {
		return nil, fmt.Errorf("打开缓存数据库失败: %v", err)
	}

	_, err = cacheDB.Exec(`
		CREATE TABLE IF NOT EXISTS data_cache (
			id INTEGER PRIMARY KEY AUTOINCREMENT,
			sensor_name TEXT NOT NULL,
			slave_id INTEGER,
			temperature REAL,
			humidity REAL,
			timestamp DATETIME DEFAULT CURRENT_TIMESTAMP,
			retry_count INTEGER DEFAULT 0
		)
	`)
	if err != nil {
		return nil, fmt.Errorf("创建缓存表失败: %v", err)
	}

	_, err = cacheDB.Exec("CREATE INDEX IF NOT EXISTS idx_cache_retry ON data_cache(retry_count)")
	if err != nil {
		return nil, fmt.Errorf("创建缓存索引失败: %v", err)
	}

	cached := &CachedStorage{
		targetStorage: targetStorage,
		cacheDBPath:   cacheDBPath,
		cacheDB:       cacheDB,
		flushInterval: flushInterval,
		batchSize:     batchSize,
		running:       true,
	}

	go cached.flushWorker()

	return cached, nil
}

func (c *CachedStorage) flushWorker() {
	ticker := time.NewTicker(time.Duration(c.flushInterval) * time.Second)
	defer ticker.Stop()

	for c.running {
		<-ticker.C
		c.flushCache()
	}
}

func (c *CachedStorage) flushCache() {
	c.mu.Lock()
	defer c.mu.Unlock()

	rows, err := c.cacheDB.Query(
		"SELECT * FROM data_cache WHERE retry_count < 10 ORDER BY id LIMIT ?",
		c.batchSize,
	)
	if err != nil {
		fmt.Printf("[缓存] 查询缓存失败: %v\n", err)
		return
	}
	defer rows.Close()

	var records []CacheRecord
	for rows.Next() {
		var record CacheRecord
		var timestamp sql.NullTime
		err := rows.Scan(
			&record.ID, &record.SensorName, &record.SlaveID,
			&record.Temperature, &record.Humidity, &timestamp, &record.RetryCount,
		)
		if err != nil {
			fmt.Printf("[缓存] 扫描记录失败: %v\n", err)
			continue
		}
		if timestamp.Valid {
			record.Timestamp = timestamp.Time
		}
		records = append(records, record)
	}

	if len(records) == 0 {
		return
	}

	fmt.Printf("[缓存] 发现 %d 条待同步数据\n", len(records))

	for _, record := range records {
		sensorRecord := &SensorRecord{
			SensorName:  record.SensorName,
			SlaveID:     record.SlaveID,
			Temperature: record.Temperature,
			Humidity:    record.Humidity,
			Timestamp:   time.Now(),
		}

		err := c.targetStorage.Save(sensorRecord)
		if err != nil {
			_, updateErr := c.cacheDB.Exec(
				"UPDATE data_cache SET retry_count = retry_count + 1 WHERE id = ?",
				record.ID,
			)
			if updateErr != nil {
				fmt.Printf("[缓存] 更新重试次数失败: %v\n", updateErr)
			}
			fmt.Printf("[缓存] 同步失败: %v\n", err)
		} else {
			_, deleteErr := c.cacheDB.Exec("DELETE FROM data_cache WHERE id = ?", record.ID)
			if deleteErr != nil {
				fmt.Printf("[缓存] 删除已同步记录失败: %v\n", deleteErr)
			}
			fmt.Printf("[缓存] 同步成功: %s\n", record.SensorName)
		}
	}
}

func (c *CachedStorage) Save(record *SensorRecord) error {
	c.mu.Lock()
	defer c.mu.Unlock()

	_, err := c.cacheDB.Exec(
		"INSERT INTO data_cache (sensor_name, slave_id, temperature, humidity) VALUES (?, ?, ?, ?)",
		record.SensorName, record.SlaveID, record.Temperature, record.Humidity,
	)
	if err != nil {
		return fmt.Errorf("保存到缓存失败: %v", err)
	}

	fmt.Printf("[缓存] 数据已保存到本地缓存: %s - 温度%.1f, 湿度%.1f\n",
		record.SensorName, record.Temperature, record.Humidity)

	err = c.targetStorage.Save(record)
	if err != nil {
		fmt.Printf("[缓存] 实时同步失败，将稍后重试: %v\n", err)
		return nil
	}

	_, err = c.cacheDB.Exec(
		"DELETE FROM data_cache WHERE sensor_name = ? AND slave_id = ? AND temperature = ? AND humidity = ? AND retry_count = 0",
		record.SensorName, record.SlaveID, record.Temperature, record.Humidity,
	)
	if err != nil {
		fmt.Printf("[缓存] 删除实时同步记录失败: %v\n", err)
	}

	fmt.Printf("[缓存] 实时同步成功: %s\n", record.SensorName)
	return nil
}

func (c *CachedStorage) SaveBatch(records []*SensorRecord) error {
	c.mu.Lock()
	defer c.mu.Unlock()

	for _, record := range records {
		_, err := c.cacheDB.Exec(
			"INSERT INTO data_cache (sensor_name, slave_id, temperature, humidity) VALUES (?, ?, ?, ?)",
			record.SensorName, record.SlaveID, record.Temperature, record.Humidity,
		)
		if err != nil {
			return fmt.Errorf("批量保存到缓存失败: %v", err)
		}
	}

	fmt.Printf("[缓存] 批量数据已保存到本地缓存，共 %d 条\n", len(records))

	for _, record := range records {
		err := c.targetStorage.Save(record)
		if err != nil {
			fmt.Printf("[缓存] 批量实时同步失败: %v\n", err)
		} else {
			_, execErr := c.cacheDB.Exec(
				"DELETE FROM data_cache WHERE sensor_name = ? AND slave_id = ? AND temperature = ? AND humidity = ? AND retry_count = 0",
				record.SensorName, record.SlaveID, record.Temperature, record.Humidity,
			)
			if execErr != nil {
				fmt.Printf("[缓存] 删除批量实时同步记录失败: %v\n", execErr)
			}
		}
	}

	return nil
}

func (c *CachedStorage) Close() error {
	c.running = false

	c.mu.Lock()
	defer c.mu.Unlock()

	c.flushCache()

	if c.cacheDB != nil {
		return c.cacheDB.Close()
	}
	return nil
}

func (c *CachedStorage) GetCacheStats() (map[string]interface{}, error) {
	c.mu.Lock()
	defer c.mu.Unlock()

	var totalCount, pendingCount int
	err := c.cacheDB.QueryRow("SELECT COUNT(*) FROM data_cache").Scan(&totalCount)
	if err != nil {
		return nil, err
	}

	err = c.cacheDB.QueryRow("SELECT COUNT(*) FROM data_cache WHERE retry_count < 10").Scan(&pendingCount)
	if err != nil {
		return nil, err
	}

	return map[string]interface{}{
		"total_records": totalCount,
		"pending_count": pendingCount,
		"cache_db_path": c.cacheDBPath,
	}, nil
}

File: Python/src/data_storage.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
数据存储模块
支持SQLite、InfluxDB、CSV三种存储方式
支持本地缓存，确保数据库不可用时数据不丢失
"""

import csv
import sqlite3
import json
import os
import threading
import queue
import time
from datetime import datetime
from abc import ABC, abstractmethod
from typing import Dict, List, Optional, Any
from pathlib import Path


class DataStorage(ABC):
    """数据存储抽象基类"""

    @abstractmethod
    def save(self, sensor_name: str, slave_id: int, temperature: float, humidity: float):
        pass

    @abstractmethod
    def save_batch(self, data: List[Dict[str, Any]]):
        pass

    @abstractmethod
    def close(self):
        pass


class CachedStorage(DataStorage):
    """带本地缓存的存储，确保数据不丢失"""

    def __init__(
        self,
        target_storage: DataStorage,
        cache_db_path: str = "sensor_cache.db",
        flush_interval: int = 10,
        batch_size: int = 100
    ):
        self.target_storage = target_storage
        self.cache_db_path = cache_db_path
        self.flush_interval = flush_interval
        self.batch_size = batch_size
        self.running = True
        self.cache_queue = queue.Queue()
        self.lock = threading.Lock()

        self._init_cache_db()
        self._start_flush_thread()

    def _init_cache_db(self):
        self.cache_conn = sqlite3.connect(self.cache_db_path, check_same_thread=False)
        self.cache_conn.row_factory = sqlite3.Row
        cursor = self.cache_conn.cursor()
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS data_cache (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                sensor_name TEXT NOT NULL,
                slave_id INTEGER,
                temperature REAL,
                humidity REAL,
                timestamp DATETIME DEFAULT CURRENT_TIMESTAMP,
                retry_count INTEGER DEFAULT 0
            )
        ''')
        cursor.execute('CREATE INDEX IF NOT EXISTS idx_processed ON data_cache(retry_count)')
        self.cache_conn.commit()

    def _start_flush_thread(self):
        self.flush_thread = threading.Thread(target=self._flush_worker, daemon=True)
        self.flush_thread.start()

    def _flush_worker(self):
        while self.running:
            try:
                time.sleep(self.flush_interval)
                self._flush_cache()
            except Exception as e:
                print(f"[缓存] 刷新线程错误: {e}")

    def _flush_cache(self):
        with self.lock:
            try:
                cursor = self.cache_conn.cursor()
                cursor.execute(
                    'SELECT * FROM data_cache WHERE retry_count < 10 ORDER BY id LIMIT ?',
                    (self.batch_size,)
                )
                records = cursor.fetchall()

                if not records:
                    return

                print(f"[缓存] 发现 {len(records)} 条待同步数据")

                for row in records:
                    try:
                        record = {
                            'sensor_name': row['sensor_name'],
                            'slave_id': row['slave_id'],
                            'temperature': row['temperature'],
                            'humidity': row['humidity'],
                            'timestamp': datetime.now()
                        }
                        self.target_storage.save(
                            record['sensor_name'],
                            record['slave_id'],
                            record['temperature'],
                            record['humidity']
                        )
                        cursor.execute('DELETE FROM data_cache WHERE id = ?', (row['id'],))
                        print(f"[缓存] 同步成功: {record['sensor_name']}")
                    except Exception as e:
                        cursor.execute(
                            'UPDATE data_cache SET retry_count = retry_count + 1 WHERE id = ?',
                            (row['id'],)
                        )
                        print(f"[缓存] 同步失败: {e}")

                self.cache_conn.commit()
            except Exception as e:
                print(f"[缓存] 刷新缓存失败: {e}")

    def save(self, sensor_name: str, slave_id: int, temperature: float, humidity: float):
        try:
            cursor = self.cache_conn.cursor()
            cursor.execute(
                'INSERT INTO data_cache (sensor_name, slave_id, temperature, humidity) VALUES (?, ?, ?, ?)',
                (sensor_name, slave_id, temperature, humidity)
            )
            self.cache_conn.commit()
            print(f"[缓存] 数据已保存到本地缓存: {sensor_name} - 温度{temperature}, 湿度{humidity}")

            try:
                self.target_storage.save(sensor_name, slave_id, temperature, humidity)
                cursor.execute(
                    'DELETE FROM data_cache WHERE sensor_name = ? AND slave_id = ? AND temperature = ? AND humidity = ? AND retry_count = 0',
                    (sensor_name, slave_id, temperature, humidity)
                )
                self.cache_conn.commit()
                print(f"[缓存] 实时同步成功: {sensor_name}")
            except Exception as e:
                print(f"[缓存] 实时同步失败，将稍后重试: {e}")

        except Exception as e:
            print(f"[缓存] 保存到缓存失败: {e}")

    def save_batch(self, data: List[Dict[str, Any]]):
        with self.lock:
            try:
                cursor = self.cache_conn.cursor()
                for item in data:
                    cursor.execute(
                        'INSERT INTO data_cache (sensor_name, slave_id, temperature, humidity) VALUES (?, ?, ?, ?)',
                        (item['sensor_name'], item['slave_id'], item['temperature'], item['humidity'])
                    )
                self.cache_conn.commit()
                print(f"[缓存] 批量数据已保存到本地缓存: {len(data)} 条")

                self._flush_cache()

            except Exception as e:
                print(f"[缓存] 批量保存到缓存失败: {e}")

    def get_cache_stats(self) -> Dict[str, Any]:
        cursor = self.cache_conn.cursor()
        cursor.execute('SELECT COUNT(*) as count FROM data_cache')
        total = cursor.fetchone()['count']
        cursor.execute('SELECT COUNT(*) as count FROM data_cache WHERE retry_count > 0')
        failed = cursor.fetchone()['count']
        return {
            'total_cached': total,
            'failed_count': failed,
            'cache_db': self.cache_db_path
        }

    def sync_all(self):
        print("[缓存] 手动触发全量同步...")
        self._flush_cache()

    def close(self):
        self.running = False
        self._flush_cache()
        if self.cache_conn:
            self.cache_conn.close()
        if self.target_storage:
            self.target_storage.close()


class SQLiteStorage(DataStorage):
    """SQLite存储"""

    def __init__(self, db_path: str = "sensor_data.db"):
        self.db_path = db_path
        self.conn = sqlite3.connect(db_path, check_same_thread=False)
        self.conn.row_factory = sqlite3.Row
        self._create_table()

    def _create_table(self):
        cursor = self.conn.cursor()
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS sensor_data (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                timestamp DATETIME DEFAULT CURRENT_TIMESTAMP,
                sensor_name TEXT NOT NULL,
                slave_id INTEGER,
                temperature REAL,
                humidity REAL
            )
        ''')
        cursor.execute('CREATE INDEX IF NOT EXISTS idx_timestamp ON sensor_data(timestamp)')
        cursor.execute('CREATE INDEX IF NOT EXISTS idx_sensor_name ON sensor_data(sensor_name)')
        self.conn.commit()

    def save(self, sensor_name: str, slave_id: int, temperature: float, humidity: float):
        cursor = self.conn.cursor()
        cursor.execute(
            'INSERT INTO sensor_data (sensor_name, slave_id, temperature, humidity) VALUES (?, ?, ?, ?)',
            (sensor_name, slave_id, temperature, humidity)
        )
        self.conn.commit()

    def save_batch(self, data: List[Dict[str, Any]]):
        cursor = self.conn.cursor()
        cursor.executemany(
            'INSERT INTO sensor_data (sensor_name, slave_id, temperature, humidity) VALUES (?, ?, ?, ?)',
            [(item['sensor_name'], item['slave_id'], item['temperature'], item['humidity']) for item in data]
        )
        self.conn.commit()

    def query(self, limit: int = 100, offset: int = 0) -> List[sqlite3.Row]:
        cursor = self.conn.cursor()
        cursor.execute('SELECT * FROM sensor_data ORDER BY timestamp DESC LIMIT ? OFFSET ?', (limit, offset))
        return cursor.fetchall()

    def query_by_time_range(self, start_time: str, end_time: str) -> List[sqlite3.Row]:
        cursor = self.conn.cursor()
        cursor.execute(
            'SELECT * FROM sensor_data WHERE timestamp BETWEEN ? AND ? ORDER BY timestamp',
            (start_time, end_time)
        )
        return cursor.fetchall()

    def query_by_sensor(self, sensor_name: str, limit: int = 100) -> List[sqlite3.Row]:
        cursor = self.conn.cursor()
        cursor.execute(
            'SELECT * FROM sensor_data WHERE sensor_name = ? ORDER BY timestamp DESC LIMIT ?',
            (sensor_name, limit)
        )
        return cursor.fetchall()

    def get_stats(self) -> Dict[str, Any]:
        cursor = self.conn.cursor()
        cursor.execute('SELECT COUNT(*) as count FROM sensor_data')
        count = cursor.fetchone()['count']
        cursor.execute('SELECT MIN(timestamp) as earliest, MAX(timestamp) as latest FROM sensor_data')
        time_range = cursor.fetchone()
        return {
            'total_records': count,
            'earliest': time_range['earliest'],
            'latest': time_range['latest']
        }

    def close(self):
        if self.conn:
            self.conn.close()


class InfluxDBStorage(DataStorage):
    """InfluxDB时序数据库存储"""

    def __init__(
        self,
        url: str = "http://localhost:8086",
        token: str = "my-token",
        org: str = "my-org",
        bucket: str = "sensor-data"
    ):
        self.url = url
        self.token = token
        self.org = org
        self.bucket = bucket
        self.client = None
        self.write_api = None
        self._connect()

    def _connect(self):
        try:
            from influxdb_client import InfluxDBClient
            from influxdb_client.client.write_api import SYNCHRONOUS

            self.client = InfluxDBClient(url=self.url, token=self.token, org=self.org)
            self.write_api = self.client.write_api(write_options=SYNCHRONOUS)
        except ImportError:
            print("警告: influxdb-client未安装，请运行: pip install influxdb-client")

    def save(self, sensor_name: str, slave_id: int, temperature: float, humidity: float):
        if not self.write_api:
            raise Exception("InfluxDB客户端未初始化")

        try:
            from influxdb_client import Point

            point = (
                Point("sensor_data")
                .tag("sensor", sensor_name)
                .tag("slave_id", str(slave_id))
                .field("temperature", temperature)
                .field("humidity", humidity)
            )
            self.write_api.write(bucket=self.bucket, org=self.org, record=point)
        except Exception as e:
            raise Exception(f"InfluxDB写入失败: {e}")

    def save_batch(self, data: List[Dict[str, Any]]):
        if not self.write_api:
            raise Exception("InfluxDB客户端未初始化")

        try:
            from influxdb_client import Point

            points = []
            for item in data:
                points.append(
                    Point("sensor_data")
                    .tag("sensor", item['sensor_name'])
                    .tag("slave_id", str(item['slave_id']))
                    .field("temperature", item['temperature'])
                    .field("humidity", item['humidity'])
                )
            self.write_api.write(bucket=self.bucket, org=self.org, record=points)
        except Exception as e:
            raise Exception(f"InfluxDB批量写入失败: {e}")

    def close(self):
        if self.client:
            self.client.close()


class CSVStorage(DataStorage):
    """CSV文件存储"""

    def __init__(self, file_path: str = None):
        if file_path is None:
            file_path = f"sensor_data_{datetime.now().strftime('%Y%m%d')}.csv"
        self.file_path = file_path
        self._create_file()

    def _create_file(self):
        if not os.path.exists(self.file_path):
            with open(self.file_path, 'w', newline='', encoding='utf-8') as f:
                writer = csv.writer(f)
                writer.writerow(['timestamp', 'sensor_name', 'slave_id', 'temperature', 'humidity'])

    def save(self, sensor_name: str, slave_id: int, temperature: float, humidity: float):
        timestamp = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
        with open(self.file_path, 'a', newline='', encoding='utf-8') as f:
            writer = csv.writer(f)
            writer.writerow([timestamp, sensor_name, slave_id, temperature, humidity])

    def save_batch(self, data: List[Dict[str, Any]]):
        timestamp = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
        with open(self.file_path, 'a', newline='', encoding='utf-8') as f:
            writer = csv.writer(f)
            for item in data:
                writer.writerow([timestamp, item['sensor_name'], item['slave_id'],
                                item['temperature'], item['humidity']])

    def query(self, limit: int = 100) -> List[Dict[str, Any]]:
        results = []
        with open(self.file_path, 'r', encoding='utf-8') as f:
            reader = csv.DictReader(f)
            for row in reader:
                results.append(row)
                if len(results) >= limit:
                    break
        return results

    def get_stats(self) -> Dict[str, Any]:
        if not os.path.exists(self.file_path):
            return {'total_records': 0}
        with open(self.file_path, 'r', encoding='utf-8') as f:
            reader = csv.reader(f)
            next(reader)
            count = sum(1 for _ in reader)
        return {'total_records': count}

    def close(self):
        pass


class StorageFactory:
    """存储工厂类"""

    STORAGE_TYPES = {
        'sqlite': SQLiteStorage,
        'influxdb': InfluxDBStorage,
        'csv': CSVStorage,
        'none': type(None)
    }

    @staticmethod
    def create_storage(storage_type: str, use_cache: bool = True, **kwargs) -> DataStorage:
        """
        创建存储实例

        :param storage_type: 存储类型 ('sqlite', 'influxdb', 'csv', 'none')
        :param use_cache: 是否使用本地缓存确保数据不丢失
        :param kwargs: 存储配置参数
        :return: 存储实例
        """
        storage_type = storage_type.lower()

        if storage_type == 'none' or storage_type == 'disabled':
            return None

        if storage_type not in StorageFactory.STORAGE_TYPES:
            print(f"警告: 未知的存储类型 '{storage_type}'，使用SQLite")
            storage_type = 'sqlite'

        storage_class = StorageFactory.STORAGE_TYPES[storage_type]
        target_storage = storage_class(**kwargs)

        if use_cache and storage_type != 'csv':
            print(f"[存储] 启用本地缓存保护，目标类型: {storage_type}")
            cache_db = kwargs.get('cache_db_path', 'sensor_cache.db')
            return CachedStorage(target_storage, cache_db_path=cache_db)

        return target_storage

    @staticmethod
    def get_available_types() -> List[str]:
        """获取可用的存储类型"""
        return list(StorageFactory.STORAGE_TYPES.keys())


def get_storage_config() -> Dict[str, Any]:
    """获取默认存储配置"""
    return {
        'type': 'sqlite',
        'use_cache': True,
        'sqlite': {
            'db_path': 'sensor_data.db'
        },
        'influxdb': {
            'url': 'http://localhost:8086',
            'token': 'my-token',
            'org': 'my-org',
            'bucket': 'sensor-data'
        },
        'csv': {
            'file_path': None
        }
    }
